Windows msvc

Rantanen opened the pull request:
- Instead of disabling Luv/Busted, we can fix the build by specifying LUAJIT in cmake files for MSVC
- The gperf issue can be fixed by demanding MSVC to use functions for ceil/floor instead of inline intrinsics
- Also tweaked the build_msvc.bat a bit:
  - Skipped gperf download if it already exists.
  - Enabled specifying cmake options (such as PYTHON_EXECUTABLE) from command line through the bat.
  - Forced GYP (libuv) to use VS2015.
  - ... and switched CMake to use VS2015 generator instead of NMake.

That last one was just personal preference. Felt more in control with solution files that could be opened in Visual Studio if needed.

Feel free to omit stuff/cherry pick changes you want.

equalsraf committed:
FIXME: third-party: Build libvterm in Windows

Using equalsraf/libvterm-win.

equalsraf committed:
third-party: Fix libuv recipe for Windows/NMake X86_64

The 64bit check for the libuv recipe worked for the VS generator
but not for NMake.

equalsraf committed:
third-party: Add helper to detect architecture

equalsraf committed:
MSVC: Include os_defs.h in kvec.h

kvec.h uses "restrict" which is not defined in MSVC, but there is a definition
in win_defs.h

Commit:
Appveyor: Enable MSVC builds

Rantanen committed:
MSVC: Fix Luv build recipe

Rantanen committed:
MSVC: Fix use of ceil/floor in gperf

MSVC uses intrinsic inline ceil and floor by default. This prevents gperf from
acquiring function pointers for these two functions.

Declaring `#pragma function(floor, ceil)` forces real functions to be used.

equalsraf commented on .ci/build_msvc.bat:
```diff
@@ -1,19 +1,34 @@
-C:\msys64\usr\bin\bash -lc "pacman --verbose --noconfirm -Su" || goto :error
-C:\msys64\usr\bin\bash -lc "pacman --verbose --noconfirm --needed -S gperf" || goto :error
 
+set CMAKE_GENERATOR="Visual Studio 14 2015 Win64"
```
My olny reason for using nmake is that it is usually more verbose on errors. With the VS generator the vcvarsall line can be removed, I think.

equalsraf commented on src/nvim/eval.c:
```diff
@@ -6706,6 +6706,11 @@ static int get_env_tv(char_u **arg, typval_T *rettv, int evaluate)
+// funcs.generated.h requires function forms for function pointers.
+# ifdef _MSC_VER
+#  pragma function(ceil, floor)
+# endif
```
@Rantanen nice work cathing this. Can this be in win_defs.h instead?

equalsraf commented:
Awesome, thanks.

Rantanen committed:
Move #pragma function to win_defs.h

Commit:
debug builds

equalsraf commented:
@Rantanen I've added most of your changes, also enabled the functional tests. Some tests are failing, this seems to be caused by the missing libintl.

Log: https://ci.appveyor.com/project/equalsraf/neovim/build/1373/job/avdu5wkw4hvo2b1v
Source: https://github.com/neovim/neovim/blob/51937e1322de0e4348c217947c117f54f3c664c3/src/nvim/ex_docmd.c#L141

Rantanen committed:
Get gperf conditionally in MSVC builds.

Rantanen committed:
Pass command line args to CMake in MSVC builds

Allows developers to specify options such as -DPYTHON_EXECUTABLE from command
line when invoking the batch files.

Rantanen commented:
Oh, the tests are working? That's neat! Last I checked nvim.exe wouldn't even start so I guess this might mean it might.

I tested switching the `#pragma` to `win_defs.h` and that seemed to work fine.

Also rebased some of the remaining changes to the `build_msvc.bat`. These aren't needed for the CI builds, but would help if the batch file is being used by other developers when building on Windows.

Tell me if you want to keep the `build_msvc.bat` more deterministic and aimed purely for the CI-environment and I'll figure out some other way to allow easy dev-builds with the MSVC toolchain. Currently I've been using the `build_msvc.bat` for this. :)

equalsraf commented:
> Oh, the tests are working? That's neat! Last I checked nvim.exe wouldn't even start so I guess this might mean it might.

A lot more tests are failing in debug builds
- https://ci.appveyor.com/project/equalsraf/neovim/build/1374/job/m36fsroanp6psa0t

most of them seem to be shada related.

> Tell me if you want to keep the build_msvc.bat more deterministic and aimed purely for the CI

Here its fine. For merging upstream its probably best to leave it out, since it is in a separate commit this is easy to do in a rebase.

I would leave out the gperf check, we need to do some CI fixes about dependenceies anyway, so this can be done later on.

Rantanen commented:
> I would leave out the gperf check, we need to do some CI fixes about dependenceies anyway, so this can be done later on.

I'll just keep it in my branch then as I don't have msys installed locally. So don't be surprised if it pops up somewhere else as well - just ignore it when picking commits. :)

equalsraf commented:
> I'll just keep it in my branch then as I don't have msys installed locally.

In Unix third-party builds gperf. In Windows with MSVC we could download a binay build instead.

Rantanen commented:
> we could download a binary build instead.

That makes sense.

equalsraf commented:
Fixed gperf recipe to grab binary builds from gnuwin32

Rantanen commented:
Tried looking into the libintl/iconv problem today. Given those are C libraries, we should be able to crosslink MinGW built libraries into MSVC build. Unfortunately I couldn't find 64-bit prebuilt binaries. The Gnuwin32 binaries are built for x86.

Rantanen commented:
Oh, also closing this pull request.

All the changes that were included here are now either taken into the windows-msvc branch or implemented some other way.
